In [13]:
import os 
import sys
module_path = "compiled_python_modules/"
sys.path.append(module_path)
import refolding_many_temperatures

import numpy as np
import matplotlib.pyplot as plt
from numba import njit

from general_python_module import *

### Load relevant variables

In [14]:
MJ_tab_6 = load_MJ_tab_6() # conventional order of amino acids given by num_to_letter_map variable
hamiltonian_walks_contacts=load_hamiltonian_walks_contacts()
hamiltonian_walks_max_abs_GE=load_hamiltonian_walks_max_abs_GE()
hamiltonian_walks_i1_i2=load_hamiltonian_walks_maGE_loop_i1_i2()
hamiltonian_walks_maGE_thread_is_N=load_hamiltonian_walks_maGE_thread_is_N()


### Plot the distribution of the Gaussian enanglement
One of the loaded variables, `hamiltonian_walks_max_abs_GE`, contains the calculated value of maximum absolute Gaussian enanglement of each of the 103346 Hamiltonain walks on the 3x3x3 cubic lattice.
 
The following cell plots the histogram of such values.

In [15]:
i_struct=897
i_seq=0

print(f'The GE of structure {i_struct:} is {hamiltonian_walks_max_abs_GE[i_struct-1]:4.3f}.')

native_contacts=hamiltonian_walks_contacts[i_struct-1]
number_of_native_contacts=len(native_contacts)

print(f"Structure {i_struct:} is characterized by these contacts:")
print(f"{native_contacts[0]:} {native_contacts[1]:} ... {native_contacts[-2]:} {native_contacts[-1]:}")

numeric_msa=np.loadtxt(f"data/sequences_data/z_score_opt_sequences/struct_{i_struct:06d}/numeric_msa.txt",dtype=np.int32)
sequence=numeric_msa[0,:]
letters_sequence="".join([num_to_letter_map[v-1] for v in sequence])

print(f'The selected sequence is {letters_sequence:}.')



The GE of structure 897 is 0.700.
Structure 897 is characterized by these contacts:
[1 6] [ 1 22] ... [17 20] [18 27]
The selected sequence is WQKGDMMKGGDWWKAVKGTDYMLAMSG.


In [16]:
import csv
with open(os.path.join("data/structural_and_entanglement_data/","loops_indexes.csv"), newline='') as f:
    reader = csv.reader(f)
    tmp_all_loops_indexes = list(reader)
tmp_all_loops_indexes=[[int(x) for x in sublist] for sublist in tmp_all_loops_indexes]
all_loops_indexes=[]
for row in tmp_all_loops_indexes:
    all_loops_indexes.append([])
    for ii in range(len(row)//2):
        all_loops_indexes[-1].append([row[2*ii],row[2*ii+1]])
del tmp_all_loops_indexes

with open(os.path.join("data/structural_and_entanglement_data/","N_threads_indexes.csv"), newline='') as f:
    reader = csv.reader(f)
    tmp_all_N_threads_indexes = list(reader)
tmp_all_N_threads_indexes=[[int(x) for x in sublist] for sublist in tmp_all_N_threads_indexes]
all_N_threads_indexes=[]
for row in tmp_all_N_threads_indexes:
    all_N_threads_indexes.append([])
    for ii in range(len(row)//2):
        all_N_threads_indexes[-1].append([row[2*ii],row[2*ii+1]])
del tmp_all_N_threads_indexes

with open(os.path.join("data/structural_and_entanglement_data/","C_threads_indexes.csv"), newline='') as f:
    reader = csv.reader(f)
    tmp_all_C_threads_indexes = list(reader)
tmp_all_C_threads_indexes=[[int(x) for x in sublist] for sublist in tmp_all_C_threads_indexes]
all_C_threads_indexes=[]
for row in tmp_all_C_threads_indexes:
    all_C_threads_indexes.append([])
    for ii in range(len(row)//2):
        all_C_threads_indexes[-1].append([row[2*ii],row[2*ii+1]])
del tmp_all_C_threads_indexes

with open(os.path.join("data/structural_and_entanglement_data/","N_threads_GE.csv"), newline='') as f:
    reader = csv.reader(f)
    all_N_threads_GE = list(reader)
all_N_threads_GE=[[float(x) for x in sublist] for sublist in all_N_threads_GE]

with open(os.path.join("data/structural_and_entanglement_data/","C_threads_GE.csv"), newline='') as f:
    reader = csv.reader(f)
    all_C_threads_GE = list(reader)
all_C_threads_GE=[[float(x) for x in sublist] for sublist in all_C_threads_GE]

In [17]:
hamiltonian_walks_bond_vectors=load_hamiltonian_walks_bond_vectors()

native_structure_1_6 = hamiltonian_walks_bond_vectors[i_struct-1]
print('for the conformation:')
for bond in native_structure_1_6:
    print(f"{bond:} ",end='')
print()
print('I get:')
for ii in range(len(all_N_threads_GE[i_struct-1])):
    print(f"""\
loop: {all_loops_indexes[i_struct-1][ii][0]:>2} - {all_loops_indexes[i_struct-1][ii][1]:>2} \t\
G_N: {all_N_threads_GE[i_struct-1][ii]:>8.3f} \t\
G_C: {all_C_threads_GE[i_struct-1][ii]:>8.3f} \
""")

for the conformation:
1 1 3 2 2 3 1 1 5 2 2 4 1 5 4 1 3 6 4 2 2 5 3 3 1 1 
I get:
loop:  1 - 22 	G_N:      nan 	G_C:    0.084 
loop:  2 - 21 	G_N:      nan 	G_C:    0.132 
loop:  3 - 20 	G_N:      nan 	G_C:    0.183 
loop:  4 - 19 	G_N:      nan 	G_C:    0.258 
loop:  5 - 14 	G_N:   -0.045 	G_C:    0.161 
loop:  6 - 13 	G_N:   -0.026 	G_C:    0.061 
loop: 10 - 19 	G_N:   -0.024 	G_C:    0.259 
loop: 10 - 27 	G_N:   -0.001 	G_C:      nan 
loop: 11 - 26 	G_N:   -0.008 	G_C:      nan 
loop: 12 - 25 	G_N:   -0.036 	G_C:      nan 
loop: 13 - 22 	G_N:   -0.038 	G_C:    0.101 
loop: 13 - 24 	G_N:   -0.043 	G_C:      nan 
loop: 14 - 21 	G_N:   -0.029 	G_C:    0.084 
loop: 15 - 24 	G_N:    0.080 	G_C:      nan 
loop: 15 - 26 	G_N:    0.071 	G_C:      nan 
loop: 16 - 23 	G_N:    0.133 	G_C:   -0.011 
loop: 18 - 27 	G_N:    0.700 	G_C:      nan 


In [18]:
n_temperatures_to_do=11
min_temperature=0.1
max_temperature=0.3
temperatures_to_do=np.linspace(min_temperature,max_temperature,n_temperatures_to_do)
n_initial_saw_steps = 100000

n_refolding_steps=int(1e6)
n_refolding_savings=200
n_monomers=len(sequence)
n_indepentend_refolding_runs_each_temperature = 1
prng_seed=np.random.randint(10,10000000)



(   all_rep_obtained_contacts,
    all_rep_obtained_energies,
    all_rep_obtained_natives,
    all_rep_obtained_distances2,
    all_rep_obtained_bond_vectors)=\
    refolding_many_temperatures.run_refolding_simulations(
        n_refolding_steps=n_refolding_steps,
        n_refolding_savings=n_refolding_savings,
        n_monomers=n_monomers,
        n_indepentend_refolding_runs_each_temperature=n_indepentend_refolding_runs_each_temperature,
        prng_seed=prng_seed,
        interaction_matrix=MJ_tab_6,
        number_of_native_contacts=number_of_native_contacts,
        native_contacts=native_contacts,
        sequence=sequence,
        n_temperatures_to_do=n_temperatures_to_do,
        temperatures_to_do=temperatures_to_do,   
        n_initial_saw_steps=n_initial_saw_steps,
    )


In [19]:
i_snap=20
i_rep=0
i_T=8


now_conformation=all_rep_obtained_bond_vectors[i_snap,:,i_rep,i_T]
now_contacts=all_rep_obtained_contacts[i_snap,:,:,i_rep,i_T]

@njit
def calculate_GE_of_conformation(conformation,contacts):
    N = len(conformation)+1 # conformation is given as (N-1) bond vector numbers i.e. numbers [1-6]
    bonds_conv=np.array([
        [-1.0, 0.0, 0.0],
        [1.0, 0.0, 0.0],
        [0.0, -1.0, 0.0],
        [0.0, 1.0, 0.0],
        [0.0, 0.0, -1.0],
        [0.0, 0.0, 1.0],
    ])

    pos_aa=np.zeros((N,3))
    for i_aa in range(1,N):
        pos_aa[i_aa,:]=pos_aa[i_aa-1,:]+bonds_conv[conformation[i_aa-1]-1,:]
    pos_bv=0.5*(pos_aa[1:]+pos_aa[:-1])

    bv=np.zeros((N-1,3))
    for i_bv in range(N-1):
        bv[i_bv,:]=bonds_conv[conformation[i_bv]-1,:]

    # pos_bv=np.zeros((N-1,3))
    # for i_bv in range(1,N-1):
    #     pos_bv[i_bv,:]=pos_bv[i_bv-1,:]+bonds_conv[conformation[i_bv-1]-1,:]


    n_loops_to_check=len(contacts) # contacts are in 1-27 convention
    loop_indexes=[] # loop_indexes are in 0-26 convention
    for i_loop in range(n_loops_to_check):
        if contacts[i_loop,0]>0:
            if contacts[i_loop,1]-contacts[i_loop,0]>6:
                loop_indexes.append([contacts[i_loop,0]-1,contacts[i_loop,1]-1])
        else:
            break
    loop_indexes=np.array(loop_indexes)
    n_loops=len(loop_indexes)

    Nter_threads_indexes=np.zeros((n_loops),dtype=np.int32)
    Cter_threads_indexes=np.zeros((n_loops),dtype=np.int32)
    for i_loop in range(n_loops):
        if loop_indexes[i_loop,0]>3:
            Nter_threads_indexes[i_loop]=loop_indexes[i_loop,0]-2
        if loop_indexes[i_loop,1]<(N-4):
            Cter_threads_indexes[i_loop]=loop_indexes[i_loop,1]+2

    Nter_GE=np.ones((n_loops))*np.nan
    Cter_GE=np.ones((n_loops))*np.nan

    for i_loop in range(n_loops):

        N_GE=0.0
        for ii in range(Nter_threads_indexes[i_loop]):
            for jj in range(loop_indexes[i_loop,0],loop_indexes[i_loop,1]):
                Ri__Rj=pos_bv[ii]-pos_bv[jj]
                DeltaRi__times__DeltaRj=np.cross(bv[ii],bv[jj])

                now_contribution=np.dot(Ri__Rj,DeltaRi__times__DeltaRj)/(4*np.pi*((Ri__Rj**2).sum())**(3/2))
                N_GE+=now_contribution
        if Nter_threads_indexes[i_loop]>0:
            Nter_GE[i_loop]=N_GE
        
        C_GE=0.0
        for ii in range(Cter_threads_indexes[i_loop],N-1):
            for jj in range(loop_indexes[i_loop,0],loop_indexes[i_loop,1]):
                Ri__Rj=pos_bv[ii]-pos_bv[jj]
                DeltaRi__times__DeltaRj=np.cross(bv[ii],bv[jj])
                now_contribution=np.dot(Ri__Rj,DeltaRi__times__DeltaRj)/(4*np.pi*((Ri__Rj**2).sum())**(3/2))
                C_GE+=now_contribution
        if Cter_threads_indexes[i_loop]>0:
            Cter_GE[i_loop]=C_GE




    return loop_indexes,Nter_GE,Cter_GE,Nter_threads_indexes,Cter_threads_indexes


loop_indexes, Nter_GE, Cter_GE, Nter_threads_indexes, Cter_threads_indexes = calculate_GE_of_conformation(now_conformation,now_contacts)


In [20]:
def return_contacts(conformation):
    N = len(conformation)+1 # conformation is given as (N-1) bond vector numbers i.e. numbers [1-6]
    bonds_conv=np.array([
        [-1.0, 0.0, 0.0],
        [1.0, 0.0, 0.0],
        [0.0, -1.0, 0.0],
        [0.0, 1.0, 0.0],
        [0.0, 0.0, -1.0],
        [0.0, 0.0, 1.0],
    ])

    pos_aa=np.zeros((N,3))
    for i_aa in range(1,N):
        pos_aa[i_aa,:]=pos_aa[i_aa-1,:]+bonds_conv[conformation[i_aa-1]-1,:]

    contacts=[]
    for ii in range(N-2):
        for jj in range(ii+3,N):
            DR_ij=pos_aa[jj,:]-pos_aa[ii,:]
            norm_DR_ij=(DR_ij**2).sum()
            if np.isclose(norm_DR_ij,1.0):
                contacts.append([ii+1,jj+1])

    return np.array(contacts)
contacts=return_contacts(now_conformation)
contacts,now_contacts

(array([[ 1,  6],
        [ 3, 20],
        [ 5,  8],
        [ 5, 14],
        [ 6, 13],
        [ 7, 12],
        [ 8, 11],
        [11, 14],
        [17, 20]]),
 array([[ 1,  6],
        [ 3, 20],
        [ 5,  8],
        [ 5, 14],
        [ 6, 13],
        [ 7, 12],
        [ 8, 11],
        [11, 14],
        [17, 20],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        [ 0,  0],
        

In [21]:
print('for the conformation:')
for bond in now_conformation:
    print(f"{bond:} ",end='')
print()
print('I get:')
for ii in range(len(loop_indexes)):
    print(f"""\
loop: {loop_indexes[ii][0]:>2} - {loop_indexes[ii][1]:>2} \t\
G_N: {Nter_GE[ii]:>8.3f} \t\
G_C: {Cter_GE[ii]:>8.3f} \
""")

for the conformation:
1 3 5 2 4 5 3 5 2 6 4 6 3 3 6 1 6 1 5 3 3 6 6 2 2 2 
I get:
loop:  2 - 19 	G_N:      nan 	G_C:    0.028 
loop:  4 - 13 	G_N:   -0.021 	G_C:    0.003 
loop:  5 - 12 	G_N:    0.013 	G_C:    0.015 


In [11]:
comparison_conformation=np.array([1, 3, 1, 4, 4, 1, 3, 3, 5, 4, 4, 2 ,3 ,5 ,2, 3, 1, 6, 2, 4, 4, 5, 1, 1, 1, 5])
comparison_contacts=return_contacts(comparison_conformation)
loop_indexes, Nter_GE, Cter_GE, Nter_threads_indexes, Cter_threads_indexes = \
    calculate_GE_of_conformation(comparison_conformation,comparison_contacts)


print('for the conformation:')
for bond in comparison_conformation:
    print(f"{bond:} ",end='')
print()
print('I get:')
for ii in range(len(loop_indexes)):
    print(f"""\
loop: {loop_indexes[ii][0]:>2} - {loop_indexes[ii][1]:>2} \t\
G_N: {Nter_GE[ii]:>8.3f} \t\
G_C: {Cter_GE[ii]:>8.3f} \
""")

for the conformation:
1 3 1 4 4 1 3 3 5 4 4 2 3 5 2 3 1 6 2 4 4 5 1 1 1 5 
I get:
loop:  1 - 20 	G_N:      nan 	G_C:   -0.056 
loop:  2 - 19 	G_N:      nan 	G_C:   -0.104 
loop:  3 - 18 	G_N:      nan 	G_C:   -0.180 
loop:  4 - 13 	G_N:    0.023 	G_C:   -0.122 
loop:  5 - 12 	G_N:   -0.007 	G_C:   -0.024 
loop:  9 - 18 	G_N:    0.026 	G_C:   -0.133 
loop: 11 - 24 	G_N:   -0.014 	G_C:      nan 
loop: 12 - 21 	G_N:   -0.039 	G_C:   -0.058 
loop: 12 - 23 	G_N:   -0.005 	G_C:      nan 
loop: 13 - 20 	G_N:   -0.017 	G_C:   -0.036 
loop: 14 - 23 	G_N:   -0.128 	G_C:      nan 
loop: 15 - 22 	G_N:   -0.181 	G_C:    0.012 


/var/folders/n5/f496lkj93z1d5p8bbkfqkhtr0000gn/T/ipykernel_25060/796841872.py:74: RuntimeWarning: invalid value encountered in scalar divide
  now_contribution=np.dot(Ri__Rj,DeltaRi__times__DeltaRj)/(4*np.pi*((Ri__Rj**2).sum())**(3/2))


for the conformation:
[1 3 1 4 4 1 3 3 5 4 4 2 3 5 2 3 1 6 2 4 4 5 1 1 1 5]
I get:
loop:  2 - 21     G_N:        nan     G_C:     -0.056
loop:  3 - 20     G_N:        nan     G_C:     -0.104
loop:  4 - 19     G_N:        nan     G_C:     -0.180
loop:  5 - 14     G_N:        nan     G_C:     -0.122
loop:  6 - 13     G_N:        nan     G_C:     -0.024
loop: 10 - 19     G_N:      0.026     G_C:     -0.133
loop: 12 - 25     G_N:     -0.014     G_C:        nan
loop: 13 - 22     G_N:     -0.039     G_C:     -0.058
loop: 13 - 24     G_N:     -0.005     G_C:        nan
loop: 14 - 21     G_N:     -0.017     G_C:     -0.036
loop: 15 - 24     G_N:     -0.128     G_C:        nan
loop: 16 - 23     G_N:     -0.181     G_C:      0.012

In [12]:
hamiltonian_walks_bond_vectors=load_hamiltonian_walks_bond_vectors()
native_structure_1_6 = hamiltonian_walks_bond_vectors[i_struct-1]

comparison_conformation=native_structure_1_6
comparison_contacts=return_contacts(native_structure_1_6)
loop_indexes, Nter_GE, Cter_GE, Nter_threads_indexes, Cter_threads_indexes = \
    calculate_GE_of_conformation(comparison_conformation,comparison_contacts)


print('for the conformation:')
for bond in comparison_conformation:
    print(f"{bond:} ",end='')
print()
print('I get:')
for ii in range(len(loop_indexes)):
    print(f"""\
loop: {loop_indexes[ii][0]+1:>2} - {loop_indexes[ii][1]+1:>2} \t\
G_N: {Nter_GE[ii]:>8.3f} \t\
G_C: {Cter_GE[ii]:>8.3f} \
""")

for the conformation:
1 1 3 2 2 3 1 1 5 2 2 4 1 5 4 1 3 6 4 2 2 5 3 3 1 1 
I get:
loop:  1 - 22 	G_N:      nan 	G_C:    0.084 
loop:  2 - 21 	G_N:      nan 	G_C:    0.132 
loop:  3 - 20 	G_N:      nan 	G_C:    0.183 
loop:  4 - 19 	G_N:      nan 	G_C:    0.258 
loop:  5 - 14 	G_N:   -0.045 	G_C:    0.161 
loop:  6 - 13 	G_N:   -0.026 	G_C:    0.061 
loop: 10 - 19 	G_N:   -0.024 	G_C:    0.259 
loop: 10 - 27 	G_N:   -0.001 	G_C:      nan 
loop: 11 - 26 	G_N:   -0.008 	G_C:      nan 
loop: 12 - 25 	G_N:   -0.036 	G_C:      nan 
loop: 13 - 22 	G_N:   -0.038 	G_C:    0.101 
loop: 13 - 24 	G_N:   -0.043 	G_C:      nan 
loop: 14 - 21 	G_N:   -0.029 	G_C:    0.084 
loop: 15 - 24 	G_N:    0.080 	G_C:      nan 
loop: 15 - 26 	G_N:    0.071 	G_C:      nan 
loop: 16 - 23 	G_N:    0.133 	G_C:   -0.011 
loop: 18 - 27 	G_N:    0.700 	G_C:      nan 


/var/folders/n5/f496lkj93z1d5p8bbkfqkhtr0000gn/T/ipykernel_25060/796841872.py:74: RuntimeWarning: invalid value encountered in scalar divide
  now_contribution=np.dot(Ri__Rj,DeltaRi__times__DeltaRj)/(4*np.pi*((Ri__Rj**2).sum())**(3/2))


for the conformation:
1 1 3 2 2 3 1 1 5 2 2 4 1 5 4 1 3 6 4 2 2 5 3 3 1 1 
I get:
loop:  1 - 22 	G_N:      nan 	G_C:    0.084 
loop:  2 - 21 	G_N:      nan 	G_C:    0.132 
loop:  3 - 20 	G_N:      nan 	G_C:    0.183 
loop:  4 - 19 	G_N:      nan 	G_C:    0.258 
loop:  5 - 14 	G_N:   -0.045 	G_C:    0.161 
loop:  6 - 13 	G_N:   -0.026 	G_C:    0.061 
loop: 10 - 19 	G_N:   -0.024 	G_C:    0.259 
loop: 10 - 27 	G_N:   -0.001 	G_C:      nan 
loop: 11 - 26 	G_N:   -0.008 	G_C:      nan 
loop: 12 - 25 	G_N:   -0.036 	G_C:      nan 
loop: 13 - 22 	G_N:   -0.038 	G_C:    0.101 
loop: 13 - 24 	G_N:   -0.043 	G_C:      nan 
loop: 14 - 21 	G_N:   -0.029 	G_C:    0.084 
loop: 15 - 24 	G_N:    0.080 	G_C:      nan 
loop: 15 - 26 	G_N:    0.071 	G_C:      nan 
loop: 16 - 23 	G_N:    0.133 	G_C:   -0.011 
loop: 18 - 27 	G_N:    0.700 	G_C:      nan